vllm/model_executor/layers/fused_moe.py

```diff
@@ -241,6 +244,38 @@ def fused_moe(hidden_states: torch.Tensor,
     M, _ = hidden_states.shape
     E, N, _ = w1.shape
 
+    if is_hip():
+        # The MoE kernels are not yet supported on ROCm.
+        routing_weights = torch.softmax(gating_output,
+                                        dim=-1,
+                                        dtype=torch.float32)
+        topk_weights, topk_ids = torch.topk(routing_weights, topk, dim=-1)
+    else:
+        import vllm._moe_C as moe_kernels
+
+        num_tokens = hidden_states.numel() // hidden_states.shape[-1]
```

Comment:
It might also be slightly cleaner / easier to test if you refactor the code a little bit into something like this:
```python
if is_hip():
    topk_weights, topk_ids = fused_topk_torch(gating_output, topk)
else:
    topk_weights, topk_ids = fused_topk_cuda(gating_output, topk)
```

That would also address the problem that at the moment `M = num_tokens` in this function -- it is slightly strange to have two variables for the same thing :)

While at it, I would also add `assert gating_output.shape[0] == hidden_states.shape[0]` (for the number of tokens) above as well as `assert w1.shape[0] == gating_output.shape[1]` (for the number of expert) :)

Reply:
It would also be worth adding a small unit test of `topk_softmax_torch` vs. `fused_topk_softmax_cuda` in `test_moe.py` -- that will increase the maintainability of the code and make it easier to port to AMD going forward if that is deemed worthwhile.

Reply:
Hi @pcmoritz, thanks for the review. The reason I didn't make a new function for topk_softmax is the chance that AMD and NVIDIA GPUs may have completely different implementations for MoE. Specifically, it is possible that AMD GPUs will use the current PyTorch + Triton while NVIDIA GPUs use the custom kernels ported from TRT-LLM. So, I was thinking that vLLM would eventually have 2 separate `fused_moe` implementations.

The reason is,
1. The custom kernels from TRT-LLM may not be portable to AMD GPUs. For example, the topk softmax kernel in this PR uses CUB, which to my knowledge cannot be simply translated via hipification. Other TRT-LLM kernels that will be added in the future PR are also not easy to port, since they use CUB or CUTLASS.
2. The two implementations are different in how it permutes/pads the input activations. For example, the TRT-LLM's topk softmax kernel returns `token_expert_indices` which will be used in subsequent permutation ops while the PyTorch-based implementation does not use it. Furthermore, TRT-LLM's grouped GEMM does not require padding between the tokens routed to different experts, while the current Triton kernel does.

Reply:
That makes sense to me! Then maybe just unify the `M` and `num_tokens` variables and add the asserts about `gating_output` to make the function a little more readable?

Reply:
Sounds good! Added.